woocommerce/api/Abstract_Cacheable_API_Base.php

```diff
@@ -0,0 +1,228 @@
+	 * Checks whether the current request is cacheable.
+	 *
+	 * @since 5.10.10
+	 *
+	 * @return bool
+	 */
+	protected function is_request_cacheable() : bool {
+
+		if ( ! in_array( Cacheable_Request_Trait::class, class_uses( $this->get_request() ), true ) ) {
+			return false;
+		}
+
+		/**
+		 * Filters whether the API request is cacheable.
+		 *
+		 * Allows actors to disable API request caching when a request is normally cacheable. This may be useful
+		 * primarily for debugging situations.
+		 *
+		 * Note: this filter is only applied if the request is originally cacheable, in order to prevent issues when
+		 * a non-cacheable request is accidentally flagged as cacheable.
+		 *
+		 * @since 5.10.10
+		 * @param bool $is_cacheable whether the request is cacheable
+		 * @param SV_WC_API_Request $request the request instance
+		 */
+		return (bool) apply_filters( 'wc_plugin_' . $this->get_plugin()->get_id() . '_api_request_is_cacheable', true, $this->get_request() );
+	}
+
+
+	/**
+	 * Gets the cache lifetime for the current request.
+	 *
+	 * @since 5.10.10
+	 *
+	 * @return int
+	 */
+	protected function get_request_cache_lifetime() : int {
+
+		/**
+		 * Filters API request cache lifetime.
+		 *
+		 * Allows actors to override cache lifetime for cacheable API requests. This may be useful for debugging
+		 * API requests by temporarily setting short cache timeouts.
+		 *
+		 * @since 5.10.10
+		 * @param int $lifetime cache lifetime in seconds, 0 = unlimited
+		 * @param SV_WC_API_Request $request the request instance
+		 */
+		return (int) apply_filters( 'wc_plugin_' . $this->get_plugin()->get_id() . '_api_request_cache_lifetime' , $this->get_request()->get_cache_lifetime(), $this->get_request() );
+	}
+
+
+
+	/**
+	 * Determine whether the response was loaded from cache or not.
+	 *
+	 * @since 5.10.10
+	 * @return bool
+	 */
+	protected function is_response_loaded_from_cache() : bool {
+
+		return $this->response_loaded_from_cache;
+	}
+
+
+	/**
+	 * Loads the response for the current request from the cache, if available.
+	 *
+	 * @since 5.10.10
+	 *
+	 * @return array|null
+	 */
+	protected function load_response_from_cache() {
+
+		return get_transient( $this->get_request_transient_key() );
+	}
+
+
+	/**
+	 * Saves the response to cache.
+	 *
+	 * @since 5.10.10
+	 *
+	 * @param array $response
+	 */
+	protected function save_response_to_cache( array $response ) {
+
+		set_transient( $this->get_request_transient_key(), $response, $this->get_request_cache_lifetime());
+	}
+
+
+	/**
+	 * Gets the response data for broadcasting the request.
+	 *
+	 * Adds a flag to the response data indicating whether the response was loaded from cache.
+	 *
+	 * @since 5.10.10
+	 *
+	 * @return array
+	 */
+	protected function get_request_data_for_broadcast() : array {
+
+		$request_data = parent::get_request_data_for_broadcast();
+
+		if ( $this->is_request_cacheable() ) {
+			$request_data['force_refresh'] = $this->get_request()->should_refresh();
```

Comment:
In practice I find these two flags to be useful when a non-cached request is made, but confusing when the cache is used since they indicate that a request is actually made when it wasn't.

What if we conditionally add the `from_cache` line to this as well? So if a real request is indeed being made, the `force_refresh` and `should_cache` flags are added. But, if we're not going to be making a request and the cache response will be used, only the `from_cache` is added.

That might help reduce confusing in debugging when these requests/responses are logged as if API communication is happening when it's really not.

Reply:
> confusing when the cache is used since they indicate that a request is actually made when it wasn't.

Interesting, my goal with adding these was to _avoid_ confusion when debugging 😅. The idea was that it would be helpful to see whether these flags were properly respected - ie, we'd like to see when a request was allowed to be loaded or stored in cache in the first place. And then later, whether the response was actually loaded from cache or not.

I'm not sure I quite follow you on the "they indicate that a request is actually made" part. Do you mean the request itself should not be logged at all?

Reply:
>I'm not sure I quite follow you on the "they indicate that a request is actually made" part. Do you mean the request itself should not be logged at all?

I think logging the request is fine, but I meant that a flag noting that it wasn't a real API request is probably helpful. I'm thinking of it more as a support engineer, debugging a site that is having trouble with tax calculation and going to the logs to see what communication took place. Everything about the logged request details indicates that there was communication with the API when there may not have been. Seeing `should_cache` makes me think that a request is being made, then cached.

Reply:
> Everything about the logged request details indicates that there was communication with the API when there may not have been. Seeing `should_cache` makes me think that a request is being made, then cached.

Would you say the `from_cache` flag on the logged response is then insufficient in indicating the request was not actually made, but the response was loaded from cache?

Reply:
imo yes. It tells me something about the response, but not the request. The flags added to the request seem to conflict with that notion in my mind - if the cache was used and the request was never made, I wouldn't expect to see things like "should cache" in the logged request, because there wasn't a request made to cache.

I still approved the PR because I don't feel that strongly about it, and it doesn't need to be a blocker right now. We can tweak those messages if there is any confusion when the plugin PRs are QAed.

Reply:
Ah, I think I see your point now. Would it make sense to log the `should_cache` flag in the response data, then?